---
title: "Sequence composition of index"
output: html_notebook
---
## Introduction and background

Indexes are used to uniquely de-multiplex reads from a pool of sequenced reads. It is of utmost importance that those indexes can be distinguished. As NGS are prone to erroneous base-call, it can misidentify indexes if they look similar. Therefore the sequence of the index should be unique and be as diverse as possible. Edit distances of two string is a measurement of how many edit of a source string it takes to match it into a target string. In picking indexes sequence, one should aim for indexes combination with high enough edit distance to avoid miss-assignment

## Not just edit distance
Other than being unique and diverse, a good set of index should fulfill the sequencer requirement. For illumina sequencer, and many other, runs of bases and extreme base composition is not favorable for sequencing, due to chemistry during library construction and during sequencing. This notebook aims to devise a way to filter out indexes that do not pass these requirement.

### Loading library 
```{r}
library(stringr)
```

### Loading test data
```{r}
# test.index.seq <- c(
#   "TAAGGCGA",
#   "CGTACTAG",
#   "AGGCAGAA",
#   "TCCTGAGC",
#   "GGACTCCT",
#   "TAGGCATG",
#   "CTCTCTAC",
#   "CAGAGAGG",
#   "GCTACGCT",
#   "CGAGGCTG",
#   "AAGAGGCA",
#   "GTAGAGGA",
#   "GCTCATGA",
#   "ATCTCAGG",
#   "ACTCGCTA",
#   "GGAGCTAC",
#   "GCGTAGTA",
#   "CGGAGCCT",
#   "TACGCTGC",
#   "ATGCGCAG",
#   "TAGCGCTC",
#   "ACTGAGCG",
#   "CCTAAGAC",
#   "CGATCAGT",
#   "TGCAGCTA",
#   "TCGACGTC"
# )

```

### calculate index's base composition - GC check

```{r}
check_gc <- function(inputString){
  full.length = str_length(inputString)  
  G.count <-  str_count(inputString, "G")
  C.count <-  str_count(inputString,"C")
  return((G.count+C.count)/full.length)
}

```
### Calculate the max repeat of any base 

```{r}
check_repeat <- function(inputString) {
  
  possible_choice <- c("A", "T", "C", "G")
  full.length = str_length(inputString)
  chr.list = str_split(inputString, pattern = "", simplify = TRUE)
  max.counter =1
  for (base in possible_choice) {
    counter = 1
    for (i in seq(length(chr.list) - 1)) {
      if (chr.list[i] != chr.list[i + 1]) {
        next
      } else if(base == chr.list[i]){
        counter = counter + 1
        
      }
      if(counter >= max.counter){
        max.counter = counter
      }
    }

  }
  return(max.counter)
}

```
### Generate a list of 8-mer, representing 8bp indexes. 
```{r}

lenght_index <- 8
possible_choice <- c("A","T","C","G")

### for lack of a package to generate kmer, use recursive function
printAllKLength <- function(set, k){
  n = length(set)
  printAllKLengthRec(set,"",n,k)
}
printAllKLengthRec <- function(set,prefix,n,k){
  if (k == 0){
    cat(paste0(prefix,"\n"))
    return()
  }
  for (i in seq(1,n)){
    newPrefix = paste0(prefix,set[i])
    printAllKLengthRec(set,newPrefix,n,k-1)
  }
}

capture.output(printAllKLength(possible_choice,lenght_index), file = "eight_mer.txt", append = TRUE)

```


### Use 8mer data for filter
```{r}
eight_mer <- read.table("~/GitHub/design_index_sequence/eight_mer.txt", quote="\"", comment.char="")
eight_mer.vector <- eight_mer$V1
eight_mer.gc <- sapply(eight_mer.vector, check_gc)
eight_mer.repeat <- sapply(eight_mer.vector, check_repeat)
eight_mer.df <-data.frame(eight_mer.gc,eight_mer.repeat)
library(dplyr)
eight_mer.df %>%  filter(eight_mer.gc >0.33) %>% filter(eight_mer.gc <0.66) %>% filter(eight_mer.repeat <=2) -> eight_mer.pass.filter.df
eight_mer.pass.filter.vector <- rownames(eight_mer.pass.filter.df)

```
### Randomly sample 500 out of the pass filter
Ideally there should be more filter such that the complexity of the sequence is maximized. This sub-sampling is only to cut cost of computation.
```{r}
set.seed(92814)
eight_mer.pass.filter.vector.sample<-sample(eight_mer.pass.filter.vector,size = 500)
```

### Calculate edit distance
The target is getting a set of indexes such that within the set, the edit distance is higher than threshold. We should not focus on any particular value. Instead, we need to consider the composition of the entire set. 
```{r}
library(vwr)
library(reshape2)
library(ggplot2)       
library(viridis)

ld_mtx <- sapply(eight_mer.pass.filter.vector.sample,levenshtein.distance,targets = eight_mer.pass.filter.vector.sample)
ld.df <- melt(ld_mtx)
### trying to see what will happen by naive filtering. 
ld.df %>% filter(value>4) -> ld.filter.df
length(unique(ld.filter.df$Var2))
# ld.filter.df$value <- as.factor(ld.filter.df$value)
plot.heatmap <- ggplot(ld.filter.df,aes(Var1,Var2))
plot.heatmap + geom_tile(aes(fill=value)) +
  theme(axis.text.x = element_blank(),axis.text.y = element_blank(),axis.title.x = element_blank(),axis.title.y = element_blank())+
  scale_fill_viridis()
```
As we can see from the above simplified heatmap, we observe some white space, which are caused by value lower than threshold (ld<4) in this case. 

### Try to sample a set of index such that the minimum ld is 2 
I opt to use random sampling in here, as I do not have a better strategy than letting the computer to do trial-and-error. It is expected to run for very long time, and there may not be an answer to this. So please use try_combo_index.R on a cluster
```{r}
success_flag <- FALSE
target.index.number <- 24
### other than itself, it should not have ld < threshold
validate_pair <- function (inputList) {
  firstString = inputList[1]
  secondString = inputList[2]
  value = inputList[3]
  threshold = 2
  if (firstString != secondString) {
    if (value <= threshold) {
      return(FALSE)
    } else{
      return(TRUE)
    }
  } else{
    return(TRUE)
  }
}

trial.counter =1

while (success_flag != TRUE) {
  ## keep trying!
  sample.indexes <-
    sample(eight_mer.pass.filter.vector, size = target.index.number)
  sample_mtx <-
    sapply(sample.indexes, levenshtein.distance, targets = sample.indexes)
  sample.df <- melt(sample_mtx)
  for_flag = TRUE
  for(row in 1:nrow(sample.df)){
    if(!validate_pair(sample.df[row,])){
      for_flag=FALSE
      break
    }
  }
  for_flag
  if(for_flag){
    break
  }
  success_flag <- all(apply(sample.df, 1, validate_pair)) ### DONE: use a for loop to break if any false , more efficient
  trial.counter = trial.counter+1
}
# sample.indexes
print(trial.counter)
plot.heatmap <- ggplot(sample.df,aes(Var1,Var2))
plot.heatmap + geom_tile(aes(fill=value)) +
  theme(axis.text.x = element_blank(),axis.text.y = element_blank(),axis.title.x = element_blank(),axis.title.y = element_blank())+
  scale_fill_viridis()

saveRDS(sample.df,file="sample_df.rds")
```


